File: 03.DP动态规划/leetcode_2770_maximumJumps.py
from typing import List
from bisect import *
from sortedcontainers import SortedList

class Solution:
    def maximumJumps(self, nums: List[int], target: int) -> int:
        n = len(nums)
        stl = SortedList([(nums[-1], n-1)])
        dp = [-10**18]*n
        dp[n-1] = 0
        for i in range(n-2, -1, -1):
            x = nums[i]
            low, high = x-target, x+target
            l = stl.bisect_left((low, -1))
            r = stl.bisect_left((high+1, -1))-1
            
            for j in range(l, r+1):
                dp[i] = max(dp[i], dp[stl[j][1]]+1)
            stl.add((nums[i], i))
        return dp[0] if dp[0] > 0 else -1
    
if __name__ == "__main__":
    s = Solution()
    nums = [1,3,6,4,1,2]
    target = 2
    print(s.maximumJumps(nums, target))

File: 03.DP动态规划/leetcode_2770_maximumJumps.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

class Solution {
public:
    int maximumJumps(vector<int>& nums, int target) {
        int n = nums.size();
        vector<int> dp(n, INT_MIN);
        dp[n-1] = 0;
        for (int i=n-2; i>=0; i--){
            for (int j=i+1; j<n; j++){
                if (abs(nums[i]-nums[j]) <= target){
                    dp[i] = max(dp[i], dp[j]+1);
                }
            }
        }
        return dp[0] > 0 ? dp[0] : -1;
    }
};

int main(){
    Solution s;
    vector<int> nums = {1,3,6,4,1,2};
    int target = 2;
    cout << s.maximumJumps(nums, target) << endl;
}
